File: 11. Longest Substring Without Repeating Characters.cpp
// Topic: Longest Substring Without Repeating Characters
#include <bits/stdc++.h>
using namespace std;

int main()
{
    string s = "abcabcbb";
    vector<int> last(256, -1);
    int start = 0, maxlen = 0;
    for (int i = 0; i < s.size(); i++)
    {
        start = max(start, last[s[i]] + 1);
        maxlen = max(maxlen, i - start + 1);
        last[s[i]] = i;
    }
    cout << "Longest length: " << maxlen << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1) (alphabet size)

File: 11. Longest Substring Without Repeating Characters.py
"""
Longest Substring Without Repeating Characters
"""
def longest_unique_substring(s):
    last = {}
    start = 0
    maxlen = 0
    for i, ch in enumerate(s):
        if ch in last and last[ch] >= start:
            start = last[ch] + 1
        maxlen = max(maxlen, i - start + 1)
        last[ch] = i
    return maxlen

def main():
    print(longest_unique_substring('abcabcbb'))

if __name__ == '__main__':
    main()

# Time Complexity: O(n)
# Space Complexity: O(1)
